### PR TITLE
Implemented brewing

### PR DESCRIPTION
**Pull request by Seadragon91**
~~DO NOT MERGE YET!~~

This implements brewing.
- Brewing stand and window
- [Recipes](https://github.com/Seadragon91/cuberite/blob/brewing/Server/brewing.txt), this file only contains the regular recipes. because the ingredient for a regular and a splash potion (level II -> extended) is the same.

TODO:
 - [x] Code needs to be reviewed

Discussion can be here and in the [forum](http://forum.mc-server.org/showthread.php?tid=1983)

### CODE REVIEW

File: src/BrewingRecipe.cpp [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+
+#include <fstream>
+
+#define BREWING_RECIPE_FILE "brewing.txt"
+
+
+
+
+
+typedef std::list<cBrewingRecipe::cRecipe> RecipeList;
```
Comment (worktycho) [inline src/BrewingRecipe.cpp]: Why a list? It general the default container should be a vector.

File: src/BrewingRecipe.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+	/** Parses the recipe contained in the line, adds it to m_pState's recipes.
+	Logs a warning to the console on input error. */
+	void AddRecipeFromLine(const AString & a_Line, unsigned int a_LineNum);
+
+	/** Parses an item string, returns true if successful. */
+	bool ParseItem(const AString & a_String, cItem & a_Item);
+
+	struct sBrewingRecipeState;
+	sBrewingRecipeState * m_pState;
```
Comment (worktycho) [inline src/BrewingRecipe.h]: This should be a unique_ptr, not a raw pointer.

File: src/BrewingRecipe.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	cBrewingRecipe(void);
+	~cBrewingRecipe();
+
+	void ReloadRecipes(void);
+
+	struct cRecipe
+	{
+		cItem * Input;
+		cItem * Output;
+		cItem * Ingredient;
```
Comment (worktycho) [inline src/BrewingRecipe.h]: All these fields should be unique_ptr's

File: src/BrewingRecipe.cpp [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+
+
+
+
+void cBrewingRecipe::ReloadRecipes(void)
+{
+	ClearRecipes();
+	LOGD("Loading brewing recipes...");
+
+	std::ifstream f(BREWING_RECIPE_FILE, std::ios::in);
```
Comment (worktycho) [inline src/BrewingRecipe.cpp]: Generally this project uses our own cFile class. I don't know why, but @madmaxoft doesn't like the streams apis.

File: src/BrewingRecipe.cpp [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+		{
+			continue;
+		}
+
+		// Remove comments from the line:
+		size_t FirstCommentSymbol = ParsingLine.find('#');
+		if ((FirstCommentSymbol != AString::npos) && (FirstCommentSymbol != 0))
+		{
+			ParsingLine.erase(ParsingLine.begin() + static_cast<const long>(FirstCommentSymbol), ParsingLine.end());
+		}
```
Comment (worktycho) [inline src/BrewingRecipe.cpp]: If you move this if above the empty line check, and remove the comparison against a first char comment, then you can remove the following switch.

File: src/BrewingRecipe.cpp [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+		if (ParsingLine.empty())
+		{
+			continue;
+		}
+
+		// Remove comments from the line:
+		size_t FirstCommentSymbol = ParsingLine.find('#');
+		if ((FirstCommentSymbol != AString::npos) && (FirstCommentSymbol != 0))
+		{
+			ParsingLine.erase(ParsingLine.begin() + static_cast<const long>(FirstCommentSymbol), ParsingLine.end());
```
Comment (worktycho) [inline src/BrewingRecipe.cpp]: I think what you want here is `std::advance(ParsingLine.begin(), FirstCommentSymbol)` rather than the cast and addition.

File: src/BrewingRecipe.cpp (hunk 5 above)
Comment (Seadragon91) [inline src/BrewingRecipe.cpp]: I am not really sure how I should change it.

File: src/BrewingRecipe.cpp (hunk 6 above)
Comment (Seadragon91) [inline src/BrewingRecipe.cpp]: This shows an error: `no matching function for call to 'advance'`. I think he don't like the iterator.

File: src/BrewingRecipe.cpp (hunk 4 above)
Comment (Seadragon91) [inline src/BrewingRecipe.cpp]: I think you mean the cIniFile. The brewing.txt is not really a ini file, it's like the furnace.txt.

File: src/BrewingRecipe.cpp (hunk 5 above)
Comment (worktycho) [inline src/BrewingRecipe.cpp]:
```
// Remove comments from the line:
size_t FirstCommentSymbol = ParsingLine.find('#');
if ((FirstCommentSymbol != AString::npos))
{
	ParsingLine.erase(ParsingLine.begin() + static_cast<const long>(FirstCommentSymbol), ParsingLine.end());
}

if (ParsingLine.empty())
{
	continue;
}
```

File: src/BrewingRecipe.cpp (hunk 6 above)
Comment (worktycho) [inline src/BrewingRecipe.cpp]: My mistake, std::advance increments an lvalue. But its probably still worth using even with the extra two statements to first declare and the increment the iterator, given that the correct form of that cast is `static_cast<std::iterator_traits<AString::iterator>::difference_type(FirstCommentSymbol)`

File: src/BrewingRecipe.cpp (hunk 5 above)
Comment (madmaxoft) [inline src/BrewingRecipe.cpp]: Use `auto`, no need to `static_cast`.

File: src/BrewingRecipe.cpp (hunk 4 above)
Comment (worktycho) [inline src/BrewingRecipe.cpp]: No I meant cFile, the headers at src/OSSupport/File.h.

Comment (madmaxoft) [inline src/BrewingRecipe.cpp]: Nah, using streams for file IO is okay. I had some reservations for using them to convert between strings and numbers.

Comment (worktycho) [inline src/BrewingRecipe.cpp]: So why does cFile exist? Is it just legacy?

File: src/BrewingRecipe.cpp [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+		return nullptr;
+	}
+
+	// Check for splash potion
+	if (a_Input.m_ItemDamage & 0x4000)
+	{
+		// Search for the drinkable potion, the ingredients are the same
+		short SplashItemDamage = a_Input.m_ItemDamage - 8192;
+
+		for (RecipeList::const_iterator itr = m_pState->Recipes.begin(); itr != m_pState->Recipes.end(); ++itr)
```
Comment (worktycho) [inline src/BrewingRecipe.cpp]: Range for loop?

File: src/BrewingRecipe.cpp [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+
+
+
+
+
+const cBrewingRecipe::cRecipe * cBrewingRecipe::GetRecipeFrom(const cItem & a_Input, const cItem & a_Ingredient) const
+{
+	const cRecipe * BestRecipe = nullptr;
+
+	for (RecipeList::const_iterator itr = m_pState->Recipes.begin(); itr != m_pState->Recipes.end(); ++itr)
```
Comment (worktycho) [inline src/BrewingRecipe.cpp]: range for loop

File: src/BlockEntities/BrewingstandEntity.cpp [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+
+
+
+
+void cBrewingstandEntity::UpdateProgressBars(bool a_ForceUpdate)
+{
+	/** Sending an update every 3th tick, using a higher value lets look the progressbar ugly */
+	if (!a_ForceUpdate && (m_World->GetWorldAge() % 3 != 0))
```
Comment (worktycho) [inline src/BlockEntities/BrewingstandEntity.cpp]: This updates on wallclock time. Not sure how to fix it though.

**Comment by Seadragon91** [comment]
Okay I am not really sure, what the error is. I think it has to do with `cBrewingRecipe::cRecipe`.

File: src/BrewingRecipe.cpp (hunk 4 above)
Comment (NiLSPACE) [inline src/BrewingRecipe.cpp]: Even if cFile wouldn't be used in the cpp code it's still convenient to have it for the Lua API.

**Comment by worktycho** [comment]
I don't think it does. MSVC is complaining about a non-existent control flow  issue in a method you haven't touched. This sort of brokenness is good example of why we need CI to compile against the exact result of the merge.

**Comment by Seadragon91** [comment]
I get that in visual stuido
`Error	37	error C2280: 'std::unique_ptr<cItem,std::default_delete<_Ty>>::unique_ptr(const std::unique_ptr<_Ty,std::default_delete<_Ty>> &)' : attempting to reference a deleted function	C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\include\xmemory0	577	1	Cuberite`

I really think I forgot something...

**Comment by worktycho** [comment]
You're right, I'm used to make stopping on errors so I misread it. MSVC is still broken though as its trying to generate a copy constructor for a type with move only members. Try explicitly deleting the cRecipe copy constructor.

**Comment by tigerw** [comment]
According to [here](http://stackoverflow.com/questions/30120573/visual-studio-2013-does-not-delete-the-copy-constructor-when-a-user-defined-move), the bug is fixed in VS2015.

**Comment by Seadragon91** [comment]
@tigerw I installed visual studio 2015 and this has been fixed there. Appveyor would be needed to upgraded, before the workaround can be removed.

**Comment by Seadragon91** [comment]
@worktycho  After change to `std::unique_ptr`,  the recipe list contains only empty recipes. I debugged that and tried to find the cause, but had no success. Do you have an idea, whats wrong?

**Comment by Seadragon91** [comment]
Found it. Had to change `typedef std::vector<cBrewingRecipe::cRecipe> RecipeList;` to `typedef std::vector<std::unique_ptr<cBrewingRecipe::cRecipe>> RecipeList;` and then fixing all the errors.

**Comment by Seadragon91** [comment]
In the brewing stand I store the current recipes that has been found (can be 3) in a pointer array named `m_CurrentBrewingRecipes`. Also simplified and fixed the return of a Recipe in GetRecipeFrom.

**Comment by Seadragon91** [comment]
Hooks should be added, too. The question is now what hooks. For started brewing process, brewing completed and a hook for a not found brewing recipe?

**Comment by Seadragon91** [comment]
I think a hook for a completed brewing process should be enough.

**Comment by Seadragon91** [comment]
Added hooks:
- Brewing process is completing
- Brewing process is completed

**Comment by Seadragon91** [comment]
I could add a array that has the result item for every bottle slot and it could be accessed from lua. My question is do we want that?

**Comment by Seadragon91** [comment]
Are the hooks good?
Only thing is now, if we want access to the current brewing result from lua side.

**Comment by Seadragon91** [comment]
I fixed a few potions and added a array that contains the result items, for the brewing recipes, that can be accessed from to lua.

I think I have everything for the brewing. Only a good review would be needed.

**Comment by Seadragon91** [comment]
How should be proceed with this pull request?

**Comment by SphinxC0re** [comment]
rebase an squash onto current master, then we will see

**Comment by Seadragon91** [comment]
If no one is against a merge, then I will rebase and merge this in the next days.

**Comment by SphinxC0re** [comment]
You should rebase again

**Comment by SphinxC0re** [comment]
also @cuberite/server-contributors nayone against the merge?

**Comment by SphinxC0re** [comment]
@cuberite/maintainers

File: src/BrewingRecipes.cpp [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+
+
+
+
+void cBrewingRecipes::ReloadRecipes(void)
+{
+	ClearRecipes();
+	LOGD("Loading brewing recipes...");
+
+	std::ifstream f(BREWING_RECIPE_FILE, std::ios::in);
```
Comment (NiLSPACE) [inline src/BrewingRecipes.cpp]: Don't we use cFile for reading/writing files?

**Comment by NiLSPACE** [comment]
If not using cFile is not a problem, then I have no objection.

**Comment by Seadragon91** [comment]
We alread discussed about that [here](https://github.com/cuberite/cuberite/pull/2497#discussion_r40496296).

**Comment by NiLSPACE** [comment]
Whoops, then I have no objection :)

File: src/Root.h [hunk 11]
```diff
@@ -208,6 +210,7 @@ class cRoot
 
 	cCraftingRecipes * m_CraftingRecipes;
 	cFurnaceRecipe *   m_FurnaceRecipe;
+	cBrewingRecipes *  m_BrewingRecipes;
```
Comment (tigerw) [inline src/Root.h]: Break the cycle and make it a unique_ptr :)

File: src/ChunkMap.h [hunk 12]
```diff
@@ -246,6 +248,10 @@ class cChunkMap
 	Returns true if all block entities processed, false if the callback aborted by returning true. */
 	bool ForEachBlockEntityInChunk(int a_ChunkX, int a_ChunkZ, cBlockEntityCallback & a_Callback);  // Lua-accessible
 
+	/** Calls the callback for brewingstand in the specified chunk.
+	Returns true if all brewingstands processed, false if the callback aborted by returning true. */
+	bool ForEachBrewingstandInChunk(int a_ChunkX, int a_ChunkZ, cBrewingstandCallback & a_Callback);  // Lua-accessible
```
Comment (tigerw) [inline src/ChunkMap.h]: Do you need this? @worktycho, did you say something about just using the generic ForEachBlockEntity function?

File: src/OSSupport/File.cpp [hunk 13]
```diff
@@ -18,7 +18,7 @@
 cFile::cFile(void) :
 	m_File(nullptr)
 {
-	// Nothing needed yet
+	// Nothing needed ye
```
Comment (madmaxoft) [inline src/OSSupport/File.cpp]: Get rid of this change

File: src/BlockEntities/BrewingstandEntity.h [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+				continue;
+			}
+			Recipe = BR->GetRecipeFrom(GetSlot(i), GetSlot(bsIngredient));
+			if (Recipe != nullptr)
+			{
+				m_CurrentBrewingRecipes[i] = Recipe;
+				m_Results[i] = Recipe->Output->CopyOne();
+			}
+		}
+	}
```
Comment (madmaxoft) [inline src/BlockEntities/BrewingstandEntity.h]: The implementation should go into the cpp file.

File: src/BlockEntities/BrewingstandEntity.h [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+
+	/** Starts the brewing proccess. Will be called if the brewing stand gets loaded from the world. */
+	void ContinueBrewing(void)
+	{
+		// Continue brewing if number is greater than 0
+		if (m_TimeBrewed > 0)
+		{
+			m_IsBrewing = true;
+		}
+	}
```
Comment (madmaxoft) [inline src/BlockEntities/BrewingstandEntity.h]: The implementation should go into the cpp file.

File: src/UI/SlotArea.cpp [hunk 16]
```diff
@@ -2338,7 +2575,3 @@ cItem * cSlotAreaTemporary::GetPlayerSlots(cPlayer & a_Player)
 	}
 	return itr->second.data();
 }
-
-
-
-
```
Comment (madmaxoft) [inline src/UI/SlotArea.cpp]: Keep the final 5 empty lines

File: src/Root.h (hunk 11 above)
Comment (Seadragon91) [inline src/Root.h]: All 3? :)

File: src/UI/SlotArea.cpp (hunk 16 above)
Comment (SphinxC0re) [inline src/UI/SlotArea.cpp]: Why?

File: src/ChunkMap.h (hunk 12 above)
Comment (madmaxoft) [inline src/ChunkMap.h]: Keep it for now, we can handle that in another PR.

Comment (Seadragon91) [inline src/ChunkMap.h]: I think all the callbacks should be reworked, but that should be done in a  own pull request.

File: src/OSSupport/File.cpp (hunk 13 above)
Comment (Seadragon91) [inline src/OSSupport/File.cpp]: Ops :)

File: src/UI/SlotArea.cpp (hunk 16 above)
Comment (madmaxoft) [inline src/UI/SlotArea.cpp]: It makes MSVS jump around the file when editing near the end and starting the debug session

File: src/ChunkMap.h (hunk 12 above)
Comment (tigerw) [inline src/ChunkMap.h]: Alright, but we are committing to a documented new API, and with all of your insistence of backwards compatibility...

...that DeprecatedBindings file doesn't want to get larger :P

Comment (madmaxoft) [inline src/ChunkMap.h]: This will only be an API function, but not an actual call in `cWorld` / `cChunkMap`, the API functions will use a common template-based framework over `ForEachEntityInChunk()`

File: src/BrewingRecipes.cpp [hunk 17]
```diff
@@ -0,0 +1,282 @@
...
+
+void cBrewingRecipes::AddRecipeFromLine(const AString & a_Line, unsigned int a_LineNum)
+{
+	AString Line(a_Line);
+	Line.erase(std::remove_if(Line.begin(), Line.end(), isspace), Line.end());
+
+	short InputDamage;
+	short OutputDamage;
+
+	std::unique_ptr<cItem> InputItem(new cItem());
```
Comment (tigerw) [inline src/BrewingRecipes.cpp]: Make these a cpp14::make_unique?

File: src/BrewingRecipes.cpp [hunk 18]
```diff
@@ -0,0 +1,282 @@
...
+		}
+	}
+
+	// Check for gunpowder
+	if (a_Ingredient.m_ItemType == E_ITEM_GUNPOWDER)
+	{
+		if (a_Input.m_ItemDamage & 0x2000)
+		{
+			// Create new recipe and add it to list
+			std::unique_ptr<cItem> InputItem(new cItem());
```
Comment (tigerw) [inline src/BrewingRecipes.cpp]: make_unique?

File: src/ChunkMap.h (hunk 12 above)
Comment (tigerw) [inline src/ChunkMap.h]: Oh. Why not have authors check for themselves the block entity they want in the generic function? Its a lot of effort to remember/implement new templates for each new block entity.

Comment (madmaxoft) [inline src/ChunkMap.h]: Because the check is trivial in C++, but rather costly when done in Lua (each C++-to-Lua call having a penalty hit for each entity, then each check calls Lua-to-C++ virtual function just to check type)

File: src/Root.h (hunk 11 above)
Comment (Seadragon91) [inline src/Root.h]: Okay there are more lines, I think this should be done in own pull request. Or should I change only my line? :)

Comment (madmaxoft) [inline src/Root.h]: Change only yours, we can adapt the rest later(tm)

Comment (Seadragon91) [inline src/Root.h]: Okay this result in changing around 10 lines and still not works. Let it for a own pull request...

**Comment by Seadragon91** [comment]
Okay with this unique_ptr change? :)

**Comment by Seadragon91** [comment]
Are the changes okay? Can I squash and rebase?

File: src/Root.h [hunk 19]
```diff
@@ -208,6 +210,7 @@ class cRoot
 
 	cCraftingRecipes * m_CraftingRecipes;
 	cFurnaceRecipe *   m_FurnaceRecipe;
+	std::unique_ptr<cBrewingRecipes>  m_BrewingRecipes;
```
Comment (madmaxoft) [inline src/Root.h]: Remove the extra space - you're not aligning anymore, so a single space is enough.

Comment (Seadragon91) [inline src/Root.h]: Removed it.

**Comment by SphinxC0re** [comment]
@cuberite/core-team again, is anyone against a merge?

**Comment by tigerw** [comment]
🎵No not really 'cos oh! I think that you've found yourself a merged PR, etc. etc. 🎶

**Comment by SphinxC0re** [comment]
What?

**Comment by tigerw** [comment]
https://en.wikipedia.org/wiki/Cheerleader_(song)
